mobilecoind-dev-faucet/src/bin/main.rs

```diff
@@ -0,0 +1,353 @@
+                    )
+                })?
+                .await
+                .map_err(|err| {
+                    format!(
+                        "Balance check request for token id '{}' ended in error: {}",
+                        token_id, err
+                    )
+                })?;
+            balances.insert(*token_id, resp.balance);
+        }
+
+        let queue_depths = self.worker.get_queue_depths();
+
+        Ok(JsonFaucetStatus {
+            success: true,
+            err_str: None,
+            b58_address: self.monitor_b58_address.clone(),
+            faucet_payout_amounts: self
+                .faucet_payout_amounts
+                .iter()
+                .map(convert_balance_pair)
+                .collect(),
+            balances: balances.iter().map(convert_balance_pair).collect(),
+            queue_depths: queue_depths
+                .into_iter()
+                .map(|(token_id, depth)| (JsonU64(*token_id), JsonU64(depth as u64)))
+                .collect(),
+        })
+    }
+}
+
+/// Request payment from the faucet, and map the rust result onto json for
+/// rocket appropriately
+#[post("/", format = "json", data = "<req>")]
+async fn post(
+    state: &rocket::State<State>,
+    req: Json<JsonFaucetRequest>,
+) -> Json<JsonSubmitTxResponse> {
+    Json(match state.handle_post(&req).await {
+        Ok(resp) => resp.into(),
+        Err(err_str) => JsonSubmitTxResponse {
+            success: false,
+            err_str: Some(err_str),
+            ..Default::default()
+        },
+    })
+}
+
+/// Request status of the faucet, and map the rust result onto json for rocket
+/// apporpriately
+#[get("/status")]
+async fn status(state: &rocket::State<State>) -> Json<JsonFaucetStatus> {
+    Json(match state.handle_status().await {
+        Ok(resp) => resp,
+        Err(err_str) => JsonFaucetStatus {
+            success: false,
+            err_str: Some(err_str),
+            ..Default::default()
+        },
+    })
+}
+
+fn convert_balance_pair(pair: (&TokenId, &u64)) -> (JsonU64, JsonU64) {
+    (JsonU64(**pair.0), JsonU64(*pair.1))
+}
+
+#[rocket::main]
+async fn main() -> Result<(), rocket::Error> {
+    mc_common::setup_panic_handler();
+    let _sentry_guard = mc_common::sentry::init();
+
+    let config = Config::parse();
+
+    let (logger, _global_logger_guard) = create_app_logger(o!());
+    log::info!(
+        logger,
+        "Starting mobilecoind-dev-faucet HTTP on {}:{}, connecting to {}",
+        config.listen_host,
+        config.listen_port,
+        config.mobilecoind_uri,
+    );
+
+    let figment = rocket::Config::figment()
+        .merge(("port", config.listen_port))
+        .merge(("address", config.listen_host.clone()));
+
+    let state = State::new(&config, &logger).expect("Could not initialize");
+
+    let _rocket = rocket::custom(figment)
+        .mount("/", routes![post, status])
+        .manage(state)
+        .launch()
+        .await?;
```

Comment:
Consider extracting a helper for this Rocket/server setup, a la https://github.com/mobilecoinfoundation/mobilecoin/blob/master/fog/overseer/server/src/server.rs

Reply:
After looking at this, I guess I'm not sure what the point of this is.

If I want to test the assembly of the server and rocket, then I think one way is, in `mobilecoind-dev-faucet/tests`, I could add integration tests that start the server by running the compiled binary, and then poke it over http etc. But this will also be quite complex because you need a consensus network for this test to work. So at some point it seems easier to make this like one of the tests that runs `local_network.py`.

I guess I don't see that it's interesting to build another binary that starts another http server using rocket, with the same routes as this one, but isn't actually the same binary. So if I'm going to make a re-usable helper out of this, who is the other user that is consuming it?

I think for the fog overseer there is a difference that, the integration tests for that server do actually start a bunch of fog ingest server on the stack in the test code: 

https://github.com/mobilecoinfoundation/mobilecoin/blob/9d7f88efe181ee498c07f62767ff56aed18a032c/fog/overseer/server/tests/inactive_outstanding_key_idle_does_not_have_key.rs#L30
https://github.com/mobilecoinfoundation/mobilecoin/blob/9d7f88efe181ee498c07f62767ff56aed18a032c/fog/overseer/server/tests/utils/mod.rs#L27

But I don't think this is currently how we test the consensus network + mobilecoind. Consensus network has a ton of threads and starting 5 consensus nodes on the stack isn't really going to work. We used to have a lot of tests like that and it was pretty flaky and eventually got removed. (Maybe, we should test fog ingest clusters differently too, but our approach there seems to be working, so if it ain't broke maybe don't fix it?)

Reply:
Fair enough. I thought it might be interesting to exercise the faucet with a mock server, but I see your point that this is more akin to `local_network.py`

Reply:
I see -- I think we could make a mock of mobilecoind in a tests/ file, and test the faucet using this:

https://doc.rust-lang.org/cargo/commands/cargo-test.html#target-selection

> Binary targets are automatically built if there is an integration test or benchmark. This allows an integration test to execute the binary to exercise and test its behavior. The CARGO_BIN_EXE_<name> [environment variable](https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-crates) is set when the integration test is built so that it can use the [env macro](https://doc.rust-lang.org/std/macro.env.html) to locate the executable.

I think it has some benefit, like it's actually testing the final binary, and it doesn't have to build the rocket part again for the test